GCS_MAVLink/AP_HAL: banner to display RCOutput modes

rmackay9 opened the pull request:
This PR attempts to help users determine if they've setup the servo outputs correctly by displaying a banner (aka a send_text that appears during the parameter download) of the form, "RCOut: OneS:1-4 PWM:5-6".

![image](https://user-images.githubusercontent.com/1498098/73319467-d0d32600-427f-11ea-9acd-dc56123f082c.png)

This comes from this discussion in which some users were struggling to setup oneshot and/or dshot: https://discuss.ardupilot.org/t/oneshot125-42-blhelis/51462

This has been tested on a pixracer and a CubeBlack

rmackay9 commented:
I made a minor change to the formatting of the output for NeoPixels (changed from "NEOP" to "NeoP") but otherwise I think this is ready for review.

peterbarker commented on libraries/AP_HAL/RCOutput.cpp:
```diff
@@ -0,0 +1,46 @@
+    // convert mode to string
+    switch (out_mode) {
+    case MODE_PWM_NONE:
+    default:
```
Any reason to have `default:`?

If someone adds an output mode it would be nice for the compiler to tell them they had to add stuff to this function!

rmackay9 replied:
No good reason in this case so removed.  I sometimes add it because I've found that if the value somehow ends up outside the enum range (which is possible with parameters cast to an enum) the default is the way to catch it.  Still, this doesn't apply in this case so I've removed it in a follow up commit. txs.

peterbarker commented on libraries/AP_HAL/RCOutput.cpp:
```diff
@@ -0,0 +1,46 @@
+}
+
+// convert output mode to string.  helper function for implementation of get_output_mode_banner
+void AP_HAL::RCOutput::append_to_banner(char banner_msg[], uint8_t banner_msg_len, output_mode out_mode, uint8_t low_ch, uint8_t high_ch) const
```
This function doesn't actually know it's appending anything to anything.  It just knows that it;s been given a buffer to shove stuff into.

I can't offer anything better than `banner_string_bit_for_mode_and_channels`

rmackay9 replied:
well, not exactly actually.  If you look closely at the snprintf you'll see that both it's source and destination are banner_msg.  So it really is appending.

This appending using sprintf is actually the bit of this PR that I'm unsure about,  It seems to work but it does worry me a little to have the source and destination be the same.

peterbarker replied:
Ah.  Why not have append_to_banner return the output from `snprintf`?  The caller can keep track of where the next stuff is to go into the buffer and pass `&banner_msg[banner_msg_len]` into this function.  It will need to worry about clamping that `snprintf` output value in that case (or just whinge and break out of the loop if it runs out of space in the buffer).  `snprintf` returns the number of bytes that would have been required if it had successfully output everything.

rmackay9 replied:
@peterbarker, thanks again for the feedback.. before I go and do that work though, is there any advantage?  Is there anything dangerous or incorrect with how it is in this PR at the moment?

peterbarker replied:
@rmackay9 I don't think it dangerous or incorrect with our *current* implementation of snprintf.  However, I have looked at optimising https://github.com/ardupilot/ardupilot/blob/master/libraries/AP_HAL/utility/print_vprintf.cpp#L380 to do a memcpy rather than bytewise assignment, and this code would break then.  I've not found any standard for snprintf for covering the buffer-and-arguments-overlap case.

Just not something we've done.  We do a fair bit of passing-in-pointers-to-bits-of-a-buffer, 'though.

The current mechanism is obviously also hideously inefficient, too :-)  For a once-off call it's shrug-worthy, but it is n^2 rather than n.

rmackay9 replied:
@peterbarker, I haven't made it more efficient (probably less) but I've made it safer (I think) by taking a backup of the original string before doing the snprintf.

WickedShell replied:
The other option is to just `asprintf` a new string each time and delete the passed in previous one. This uses memory on the heap so you have to delete it, but saves you the copy aside, and with the longer status text support we just gained this would actually cope with that situation reasonably for fully documenting the assignment. I'm okay with the allocation/deletion stuff, since this would only be done once during startup.

rmackay9 replied:
I'm hopeful that what I've done is safe and simple enough that everyone finds it acceptable.  I don't think it's worth optimising too much in this function that is so rarely called.

So, if anyone objects to how it's done, please tell me clearly that you want it changed.

peterbarker replied:
I won't hold this up for that.  I don't think any of us see actual bugs here, just alternative implementations - and I don't see anyone volunteering to own new bugs :-)

rmackay9 replied:
@WickedShell, @peterbarker, cool.  really thanks very much for the reviews.  reviews are super valuable.. so really appreciate it.

peterbarker commented on libraries/AP_HAL_ChibiOS/RCOutput.cpp:
```diff
@@ -728,6 +728,53 @@ void RCOutput::set_output_mode(uint16_t mask, enum output_mode mode)
+
+    // handle simple case
+    if (!have_nonzero_modes) {
+        hal.util->snprintf(banner_msg, banner_msg_len, "RCOut: None");
```
Not sure this is informative enough.  How about "RCOut: no RC outputs defined"?

rmackay9 replied:
hmm.. not sure.. the new suggestion just leads to "RC" and "Out" being repeated.  "defined" is the only extra bit of info and I'm not sure that adds to much.  Sorry, I think I might leave it as-is.

peterbarker commented on libraries/AP_HAL_ChibiOS/RCOutput.cpp:
```diff
@@ -728,6 +728,53 @@ void RCOutput::set_output_mode(uint16_t mask, enum output_mode mode)
+    }
+
+    // write banner to banner_msg
+    hal.util->snprintf(banner_msg, banner_msg_len, "RCOut:");
```
Could just `strncpy` this :-)

rmackay9 replied:
Txs for the suggestion.. I looked into this just now and it seems I'd need to provide a length which seems error prone and makes me feel like I'm doing too much work for it.

peterbarker commented on libraries/AP_HAL_ChibiOS/RCOutput.cpp:
```diff
@@ -728,6 +728,53 @@ void RCOutput::set_output_mode(uint16_t mask, enum output_mode mode)
+    for (uint8_t k = 1; k < ARRAY_SIZE(ch_mode); k++) {
+        if (ch_mode[k-1] != ch_mode[k]) {
+            if (ch_mode[k-1] != MODE_PWM_NONE) {
+                AP_HAL::RCOutput::append_to_banner(banner_msg, banner_msg_len, ch_mode[k-1], curr_mode_lowest_ch + 1, k);
```
Do we need the namespacing here?

rmackay9 replied:
Apparently not. Removed, thanks!

peterbarker commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -3372,6 +3372,12 @@ void GCS_MAVLINK::send_banner()
         send_text(MAV_SEVERITY_INFO, "%s", sysid);
     }
+
+    // send RC output mode info if available
```
I wonder if we can find a more appropriate place/time to emit this message.

I'm a bit worried we might end up losing chunks of information if we load too much into send_banner.

rmackay9 replied:
I'm definitely open to suggestions on where it appears.  This nice thing about this location is it's done for all vehicles.

For copter a logical place would be at the end of Copter::init_rc_out().  Plane and Sub have similar functions but Rover and Tracker sadly do not.

peterbarker replied:
In a library would be better.  For the time being - it's fine as-is :-)

peterbarker commented on libraries/AP_HAL_ChibiOS/RCOutput.cpp:
```diff
@@ -728,6 +728,53 @@ void RCOutput::set_output_mode(uint16_t mask, enum output_mode mode)
+bool RCOutput::get_output_mode_banner(char banner_msg[], uint8_t banner_msg_len) const
+{
+    // create array of each channel's mode
+    output_mode ch_mode[NUM_GROUPS * ARRAY_SIZE(pwm_group::chan)] = {MODE_PWM_NONE};
```
This is only initialising the first element as `MODE_PWM_NONE` and the rest as zeroes.  Happens to be the same value, but we shouldn't rely on it.

rmackay9 replied:
I did not realise this!  I've changed it to just "= {}" which should initialise it to zero I think which is what we want.  txs!

rmackay9 commented:
@peterbarker, thanks very much for the review!!

tridge commented:
this needs to add in the chan_offset to take account of IOMCU

rmackay9 commented:
Thanks for all the help on this.  I think I've addressed all the issues and tested it on Copter (Pixracer and CubeBlack) and Plane (Pixracer only) and it seems to be working.

I ran it through valgrind on SITL but of course, SITL doesn't even display the banner so it wasn't very interesting.

rmackay9 commented:
One thing that I noticed while writing this is that there's no way for the iomcu_mode to be returned to MODE_PWM_NORMAL.  Once it's in one-shot mode or brushed mode there's no way as far as I can see to get it back to normal pwm.  This is probably a question for @tridge.

rmackay9 commented:
Here is another messy discussion with many participants trying to figure out dshot. https://discuss.ardupilot.org/t/pixhawk-4-holybro-arducopter-3-6-9-4-0-1rc3-no-control-over-two-motors-in-an-x8-with-dshot/51480

rmackay9 commented:
This is included in Copter-4.0.3-rc1
